### DISCUSSION TITLE
Save-keys not loading.

### DISCUSSION THREAD

**Discussion by OlegAndrianov** [strongswan/strongswan#1950] [Q&A] [answered]
Looked up for similar issues, did something, but still no luck 
I enabled them at compilation, I have them installed 
`/usr/lib/ipsec/plugins/libstrongswan-save-keys.so
/usr/lib/ipsec/plugins/libstrongswan-save-keys.la`
I added loading into configuration in strongswan.conf:

> charon {
  filelog {
    charon {
      # path to the log file, specify this as section name in versions prior to 5.7.0
      path = /var/log/charon.log
      # add a timestamp prefix
      time_format = %b %e %T
      # prepend connection name, simplifies grepping
      ike_name = yes
      # overwrite existing files
      append = no
      # increase default loglevel for all daemon subsystems
      default = 1
      # flush each line to disk
      flush_line = yes
IKE=4
ENC=1
NET=0
CHD=2
CFG=2
}
  }
    start-scripts {
        swanctl = swanctl --load-all --noprompt
    }
  plugins {
    save-keys {
        load =yes
      esp = yes
      ike = yes
      wireshark_keys = /home/ulvs/ipsecSessionKeys
    }
  }
}`

I disabled apparmor so it won't get in the way
 still i get only
`Oct 13 23:28:04 00[LIB] loaded plugins: charon-systemd random nonce x509 constraints pubkey pem openssl drbg curl kernel-netlink resolve socket-default vici updown eap-identity eap-dynamic eap-tls`
Please advise what I am missing here.

**Comment by tobiasbrunner** [comment, marked as answer]
While plugins that are loaded by default can be disabled/prioritized via `charon.plugins.<plugin>.load` even if `charon.load_modular` is disabled, that's not true for loading plugins that are not part of the compiled-in plugin list. So you have to enable `load_modular` (see [the docs on plugin load](https://docs.strongswan.org/docs/5.9/plugins/pluginLoad.html#_runtime_plugin_configuration) for details) and include the config snippets in `strongswan.d` to load the plugins (see e.g. [this example strongswan.conf](https://docs.strongswan.org/docs/5.9/config/strongswanDir.html)).

**Reply by OlegAndrianov** [reply]
Thank you for your answer that indeed helped. 
I updated charon.conf and now have the following in my strongswan.conf

> charon { 
    start-scripts { 
        swanctl = swanctl --load-all --noprompt 
    } 
load_modular=yes 
plugins { 
    include strongswan.d/charon/*.conf 
    save-keys { 
        load =yes 
      esp = yes 
      ike = yes 
      wireshark_keys = /home/ulvs/ipsecSessionKeys 
    } 
    revocation { 
    load = no 
    } 
  } 
} 
include strongswan.d/*.conf 

Maybe you can update the wiki page https://docs.strongswan.org/docs/5.9/plugins/save-keys.html with the guidance you gave.
